modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,454 @@
+ * can not be modified. Any write operation switches it lazily to the read-write mode. This allows
+ * to prevent the unnecessary data copying.
+ *
+ * <p>Data may be read via the InputStream API and modified via the OutputStream one. Changes done via
+ * OutputStream are visible via the InputStream even if InputStream is created before changes done.
+ *
+ * <p>InputStream and OutputStream created remain valid even if the underlying data storage changed from
+ * read-only to read-write.
+ *
+ * <p>Note however that implementation is not thread-safe.
+ */
+public class JdbcBinaryBuffer {
+    /** Byte array storing data. */
+    private byte[] arr;
+
+    /** Offset the data starts in the array. */
+    private int off;
+
+    /** The length of data. */
+    private int len;
+
+    /** Read only flag. */
+    private boolean isReadOnly;
+
+    /** Minimum buffer capacity. */
+    public static final int MIN_CAP = 256;
+
+    /**
+     * Create buffer which wraps the existing byte array and start working in the read-only mode.
+     *
+     * @param arr The byte array to be wrapped.
+     * @param off The offset to the first byte to be wrapped.
+     * @param len The length in bytes of the data to be wrapped.
+     */
+    public static JdbcBinaryBuffer createReadOnly(byte[] arr, int off, int len) {
+        return new JdbcBinaryBuffer(arr, off, len, true);
+    }
+
+    /**
+     * Create buffer which takes ownerhip of and wraps data in the existing byte array and
+     * starts working in the read-write mode.
+     *
+     * @param arr The byte array to be wrapped.
+     */
+    public static JdbcBinaryBuffer createReadWrite(byte[] arr) {
+        return new JdbcBinaryBuffer(arr, 0, arr.length, false);
+    }
+
+    /**
+     * Create empty buffer which starts working in the read-write mode.
+     */
+    public static JdbcBinaryBuffer createReadWrite() {
+        return new JdbcBinaryBuffer(new byte[MIN_CAP], 0, 0, false);
+    }
+
+    /**
+     * Private constructor.
+     *
+     * @param arr The byte array to be wrapped.
+     * @param off The offset to the first byte to be wrapped.
+     * @param len The length in bytes of the data to be wrapped.
+     * @param isReadOnly The read-only flag.
+     */
+    private JdbcBinaryBuffer(byte[] arr, int off, int len, boolean isReadOnly) {
+        this.arr = arr;
+        this.off = off;
+        this.len = len;
+        this.isReadOnly = isReadOnly;
+    }
+
+    /**
+     * Create shallow read-only copy of this buffer.
+     */
+    public JdbcBinaryBuffer shallowCopy() {
+        return new JdbcBinaryBuffer(arr, off, len, true);
+    }
+
+    /**
+     * Provide InputStream through which the data can be read starting from the
+     * begining.
+     *
+     * <p>Stream is not limited meaning that it would return any new data
+     * written to the buffer after stream creation.
+     *
+     * @return InputStream instance.
+     */
+    public InputStream getInputStream() {
```

Comment:
https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-Gettersandsetters

> Getters and setters should be implemented without is/get/set prefixes in internal APIs.

Please, check other part of your PR, also.